File: Labs/experiment3/F_bst_preorder.cpp
#include<iostream>
using namespace std;
#define int long long

int n, x;
struct node {
    int val;
    node* lchild = NULL;
    node* rchild = NULL;
};

void ist(node* root, int v) {
    node* tmp = root;
    node* add = new node;
    add->val = v;
    while (tmp != NULL) {
        if (v == tmp->val) return;
        else if (v > tmp->val) {
            if (tmp->rchild == NULL) tmp->rchild = add;
            else tmp = tmp->rchild;
        } else {
            if (tmp->lchild == NULL) tmp->lchild = add;
            else tmp = tmp->lchild;
        }
    }
}

void preorder(node* nd) {
    if (nd == NULL) {
        cout << 0 << " ";
        return;
    }
    cout << nd->val << " ";
    preorder(nd->lchild);
    preorder(nd->rchild);
}

signed main() {
    cin >> n >> x;
    node* root = new node;
    root->val = x;
    for (int i = 2; i <= n; i++) {
        cin >> x;
        ist(root, x);
    }
    preorder(root);
    return 0;
}

File: Labs/experiment3/F_bst_preorder.py
import sys

class Node:
    def __init__(self, val: int):
        self.val = val
        self.lchild: Node | None = None
        self.rchild: Node | None = None


def insert(root: Node, v: int) -> None:
    tmp = root
    add = Node(v)
    while tmp is not None:
        if v == tmp.val:
            return
        if v > tmp.val:
            if tmp.rchild is None:
                tmp.rchild = add
                return
            tmp = tmp.rchild
        else:
            if tmp.lchild is None:
                tmp.lchild = add
                return
            tmp = tmp.lchild


def preorder(nd: Node | None, out: list[str]) -> None:
    if nd is None:
        out.append("0")
        return
    out.append(str(nd.val))
    preorder(nd.lchild, out)
    preorder(nd.rchild, out)


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    x = int(data[1])
    root = Node(x)
    for i in range(2, n + 1):
        x = int(data[i])
        insert(root, x)

    out: list[str] = []
    preorder(root, out)
    sys.stdout.write(" ".join(out))


if __name__ == "__main__":
    main()
